### DISCUSSION TITLE
Question about VSG unit modify

### DISCUSSION THREAD

**Discussion by xiaoheimai95** [CURENT/andes#454] [Modeling] [answered]
Dear Dr.Cui and Jinning,
I added the current limit in the VSGOuterPIModel：

<img width="544" alt="截屏2023-05-10 17 00 02" src="https://github.com/CURENT/andes/assets/71546641/08313eee-427c-4980-ad06-7b628db201e9">

However, the simulation results does not reflect the current limit, I would like to ask what the problem is?
![5901683709359_ pic](https://github.com/CURENT/andes/assets/71546641/38d09a3c-e6bb-46cc-89ed-9f9f5de0716c)
![5911683709359_ pic](https://github.com/CURENT/andes/assets/71546641/6fb159e9-dfde-4719-ba49-09d1f84a24e4)
![5921683710153_ pic](https://github.com/CURENT/andes/assets/71546641/0f56f902-49ff-47bf-8c1d-250298e9efb1)

**Comment by jinningwang** [comment]
[Update]

There can be a simpler way to determine the effects of the current limit. You can use two systems with the same disturbance. One system has REGCV2 and the other has REGCV3.

---

In these figures, it is hard to determine whether the current limit is in effect or not. I would suggest giving a benchmark curve that compares similar variables from ``REGCV2``. Let's say, plotting out ``REGCV2.Idref`` and ``REGCV3.Idref``. This means you might also need a ``REGCV2`` device in your case.

**Comment by cuihantao** [comment, marked as answer]
Yeah, as always, you want to test it in two steps:

1. Apply it to the original model to make sure the outcome is expected. This step makes sure the parameters are good.
2. Apply the same disturbance to your modified model. Compare the outputs (currents, in this case).

**Reply by xiaoheimai95** [reply]
Actually, I have apply the same disturbance to a system with REGCV1 and it shows the same curves. So I have not upload the pics.
<img width="580" alt="截屏2023-05-11 14 14 46" src="https://github.com/CURENT/andes/assets/71546641/e44e5135-157c-42db-b809-1f072673975e">
<img width="606" alt="截屏2023-05-11 14 15 02" src="https://github.com/CURENT/andes/assets/71546641/20f227cc-609d-4693-84ca-9375ec3572db">

**Reply by xiaoheimai95** [reply]
As can be seen in the figure, `Iqref` has exceeded the set threshold of `Imax=1.2p.u.` in the time range of 1-2s, however, no current limit was triggered. So I have no idea what is wrong with my code.

**Reply by cuihantao** [reply]
For you to consider:

Do you think the plots for REGCV1 shows a stabilized system? If not, you might want to check the parameters.

You will need a correct REGCV1 to start with before testing the limiters in REGCV3.

**Reply by xiaoheimai95** [reply]
As far as I am concerned, the unstable system should trigger the limiters. 
Also, I have tested in the stable case and the curves of the two kind of generators are still the same.

**Reply by xiaoheimai95** [reply]
In the pics I have shown, for `REGCV3_2`, `Iref` is clearly over `1.2p.u` at 0-1s. With the `Piecewise` function, I calculate the unrestricted `Iref` value by setting `Iref_init`, which should limit `Iref` to 1.2 Max by comparing it to `Imax`, however this `Piecewise` function does not work. And I think that neither `Idref` nor `Iqref` triggers this mechanism either.

**Reply by cuihantao** [reply]
I'm happy to take a look if you can share the model, the data file, and a script/notebook to reproduce it.

**Reply by xiaoheimai95** [reply]
That would be great! This package contains the REGCV3 model, the SMIB data file (only the generator was replaced with the original SMIB file), and a script to run it.

**Reply by cuihantao** [reply]
Is `Imax=1.2` applied to the `0`-th converter? If so, `ss.REGCV3.Iref_y.v[0]` (the blue curve) is always below 1.2. 

I ran your code. `Iref` for the `0`-th converter never exceeded 1.2. See:

![bd7bf9ec9d74b8eee90ee4976f7af91](https://github.com/CURENT/andes/assets/6610051/f151abbc-6482-44dd-8992-85a8fb80711c)

**Reply by cuihantao** [reply]
The `Imax` for the `1`st converter in the system base is very large, due to a large `Sn`. 

You can check it by inspecting `ss.REGCV3.Imax`: 

```
>>> ss.REGCV3.Imax

NumParam: REGCV3.Imax, v=[1.2e+00 1.2e+07], vin=[1.2 1.2]
```

What you can do is either reduce the `Sn` for the 1st converter, or include the `Imax` column in the data file and set desired values.

**Reply by cuihantao** [reply]
The other error is in 

```python
        self.Imax = NumParam(default=1.2,
                             info="max current limitation",
                             tex_name='I_{max}',
                             unit='p.u.',
                             power=True  # < --
                             )

```

It should be instead `current=True`. Although when the rated voltage of the device is the same as the bus, `power=True` and `current=True` have the same effect.

**Reply by cuihantao** [reply]
Reducing `Imax` for `REGCV3_2` to take effect on an SMIB will likely cause nonconvergence due to the lack of power flow solution. You can test it on a larger system with synchronous generators. 

Try replacing one synchronous generator with a converter and test if the current limit works.

**Reply by xiaoheimai95** [reply]
Dear Dr. Cui, when I only replaced one generator into VSG and kept the traditional generator of balance node but reduced `Sn` in the meanwhile, increased the active load in the SMIB. The results showed that the current limit was triggered. Thank you very much for your help!

**Reply by cuihantao** [reply]
Great. Glad to learn that helps.

**Reply by xiaoheimai95** [reply]
Dr. Cui, there is still a small question. When I set up the SMIB system according to the above methods, although the current limitation can be triggered, there will be initialization problems. Is this normal?

<img width="770" alt="截屏2023-05-22 10 12 53" src="https://github.com/CURENT/andes/assets/71546641/5ec34292-186b-4780-b5e0-20c333f6e5ca">

**Reply by jinningwang** [reply]
The failed initialization can occur when developing a new model. In your case, it seems to be caused by unbalanced initialization equations. There are some discussions on similar topics that might be helpful for you.

> Dr. Cui, there is still a small question. When I set up the SMIB system according to the above methods, although the current limitation can be triggered, there will be initialization problems. Is this normal?
> 
> <img alt="截屏2023-05-22 10 12 53" width="770" src="https://user-images.githubusercontent.com/71546641/239786341-5ec34292-186b-4780-b5e0-20c333f6e5ca.png">

**Reply by xiaoheimai95** [reply]
Dear Jinning,
Thanks for your answering, in my code, I did modify the `PIId`、`PIIq` and `udref`、`uqref` equation by change the `PIvd_y`、 `PIvq_y`into ` Idref_y`、 ` Iqref_y` because the current limitation and it is the only change I have made in this part.

**Reply by xiaoheimai95** [reply]
Here is my log of init.
[init_debug.txt](https://github.com/CURENT/andes/files/11539498/init_debug.txt)

**Reply by cuihantao** [reply]
The other possible cause is that the initialized values calculated from the initialization equations are out of the limits you set.

You can check the variables on which your new limits are imposed, then compare the initialized values (logged in `init_debug.txt`) with the limits.
